### PR TITLE
feat!: Set Transaction Fee Asset on New Account

### PR DESCRIPTION
**Pull request by apopiak**
Use the types introduced in https://github.com/galacticcouncil/warehouse/pull/54 to set transaction fee asset on new account (and clean up when fee asset balance reaches zero or system account is reaped).

## Description
As described in #454 we want to automatically set a user's fee asset when it makes sense to do so.

We do this by implementing a [new callback introduced to orml_tokens](https://github.com/open-web3-stack/open-runtime-module-library/pull/754) that gets triggered whenever a new account is created for a `(asset_id, account_id)` combination. If the account id does not have a fee currency set and the asset id of the new account is an accepted fee currency, we set the user's fee currency to this new asset.

## Notes
The weights should be rerun because this changes the costs of creating and deleting accounts.

## Related Issue
Fixes: #454

## How Has This Been Tested?

- [x] local integration test
- [x] xcm transfer integration test

## Checklist:

- [x] I have updated the documentation if necessary.
- [x] I have added tests to cover my changes, regression test if fixing an issue.
- [x] This is a breaking change.

### CODE REVIEW

File: pallets/offchain-duster/src/mock.rs [hunk 1]
```diff
@@ -159,6 +159,8 @@ impl orml_tokens::Config for Test {
 	type ExistentialDeposits = ExistentialDeposits;
 	type OnDust = ();
 	type MaxLocks = ();
+	type OnNewTokenAccount = ();
+    type OnKilledTokenAccount = ();
```
Comment (apopiak) [inline pallets/offchain-duster/src/mock.rs:163]:
```suggestion
	type OnNewTokenAccount = ();
	type OnKilledTokenAccount = ();
```

File: Cargo.toml [hunk 2]
```diff
@@ -198,8 +198,20 @@ cumulus-relay-chain-interface = { git = "https://github.com/paritytech//cumulus"
...
-pallet-transaction-multi-payment = { git = "https://github.com/galacticcouncil//warehouse", rev = "e428fd81f84781c4508df72f095e7bed6bb58ea1"}
-pallet-asset-registry= { git = "https://github.com/galacticcouncil//warehouse", rev = "e428fd81f84781c4508df72f095e7bed6bb58ea1"}
-hydradx-traits = { git = "https://github.com/galacticcouncil//warehouse", rev = "e428fd81f84781c4508df72f095e7bed6bb58ea1"}
+pallet-price-oracle = { git = "https://github.com/galacticcouncil//warehouse", branch = "apopiak/set-tx-fee-asset-on-new-account"}
+pallet-relaychain-info = { git = "https://github.com/galacticcouncil//warehouse", branch = "apopiak/set-tx-fee-asset-on-new-account"}
+pallet-transaction-multi-payment = { git = "https://github.com/galacticcouncil//warehouse", branch = "apopiak/set-tx-fee-asset-on-new-account"}
+pallet-asset-registry= { git = "https://github.com/galacticcouncil//warehouse", branch = "apopiak/set-tx-fee-asset-on-new-account"}
+hydradx-traits = { git = "https://github.com/galacticcouncil//warehouse", branch = "apopiak/set-tx-fee-asset-on-new-account"}
```
Comment (enthusiastmartin) [inline Cargo.toml:205]: I would prefer to use a specific revision. So when it gets merged in warehouse master - no change will be needed here.

File: pallets/xyk/src/tests.rs [hunk 3]
```diff
@@ -963,7 +963,12 @@ fn discount_sell_fees_should_work() {
 
 		expect_events(vec![
 			Event::PoolCreated(user_1, asset_a, HDX, 10_000, share_token_native, native_pair_account).into(),
-			pallet_asset_registry::Event::Registered(1, bounded_name, AssetType::PoolShare(asset_a, asset_b)).into(),
+			pallet_asset_registry::Event::Registered {
+				asset_id: 1,
+				asset_name: bounded_name,
+				asset_type: AssetType::PoolShare(asset_a, asset_b),
+			}
+			.into(),
```
Comment (enthusiastmartin) [inline pallets/xyk/src/tests.rs:971]: Not part of this PR - this is due to recent changes in warehouse and it was yet updated in Basilisk.

UI needs to be aware of this change.

I wonder it would be better to do it as part of separate PR so it can be easily tracked.

**Review by enthusiastmartin — APPROVED** [review]
This looks ok.

I wonder where the update of event structs should be as part of standalone PR  so it can be easily tracked ( because UI needs to be aware ).

File: pallets/xyk/src/tests.rs (hunk 3 above)
Comment (enthusiastmartin) [reply pallets/xyk/src/tests.rs:971]: It should be done as part of #459

File: runtime/common/src/weights/tokens.rs [hunk 4]
```diff
@@ -50,28 +50,28 @@ pub struct BasiliskWeight<T>(PhantomData<T>);
 
 impl<T: frame_system::Config> WeightInfo for BasiliskWeight<T> {
 	fn transfer() -> Weight {
-		(66_923_000 as Weight)
-			.saturating_add(T::DbWeight::get().reads(4 as Weight))
-			.saturating_add(T::DbWeight::get().writes(3 as Weight))
+		(66_005_000 as Weight)
```
Comment (jak-pan) [inline runtime/common/src/weights/tokens.rs:53]: Are these extra reads and writes part of our change or is this unrelated orml thing?

Comment (enthusiastmartin) [reply runtime/common/src/weights/tokens.rs:53]: i believe it is the extra read and write of "set fee asset" handler.

File: Cargo.toml (hunk 2 above)
Comment (apopiak) [reply Cargo.toml:205]: Right, this was for development :+1:

File: runtime/common/src/weights/tokens.rs (hunk 4 above)
Comment (apopiak) [reply runtime/common/src/weights/tokens.rs:53]: Jep, that's why I asked to run the weights, cause I wanted us to see the impact.

File: pallets/xyk/src/tests.rs (hunk 3 above)
Comment (apopiak) [reply pallets/xyk/src/tests.rs:971]: I guess we'll wait for #459 before merging this?

Comment (enthusiastmartin) [reply pallets/xyk/src/tests.rs:971]: yes. it will be merged shortly.

**Comment by lumir-mrkva** [comment]
Task linked: [CU-2kpe21t Set Fee Asset Id on New Account](https://app.clickup.com/t/2kpe21t)

**Comment by apopiak** [comment]
requires https://github.com/galacticcouncil/Basilisk-node/pull/493 (and includes its changes)
